### ISSUE TITLE
[WIP] /etc/hosts and ansible_fqdn

### ISSUE THREAD

**Issue by evilham** [debops/debops#620]
Alrighty, I came across a bit of a weird issue (don't I always?).

Disclaimer: I am trying, but since my time is limited right now, I am brain-dumping and therefore this may not make any sense if you are not in my brain :-D will improve tomorrow.
I am about 90% sure this can be solved from `debops.bootstrap` would appreciate input on that.

# Background

`ansible_fqdn` comes from `python -c 'import socket; print(socket.getfqdn())'`, which is kind of messed up if your hostname only has `AAAA` records.

# Situation:
* FQDN: `host.domain.tld` with publicly resolving `AAAA` record, but no `A` record

## `/etc/hosts`
Result of `debops bootstrap` is that `/etc/hosts` looks as follows:
```
127.0.0.1       localhost
127.0.0.1       host.domain.tld  host

# The following lines are desirable for IPv6 capable hosts
::1             localhost ip6-localhost ip6-loopback
ff02::1         ip6-allnodes
ff02::2         ip6-allrouters
```

By not having a line showing `::1 host.domain.tld host`, the result of the command that generates `ansible_fqdn` is `localhost` and not `host.domain.tld`.

# Solution?
Modify `debops.bootstrap` (and potentially `debops.netbase`) so that *in addition to* the IPv4 line, we have an IPv6 line for the FQDN.

## Status
I quickly tested by manually fixing `/etc/hosts` and after that the python command and `ansible -m setup` show the right `ansible_fqdn`.
*However* `debops bootstrap` sets up `/etc/hosts` to have `127.0.0.1 localhost` twice, i.e. it is recognising `ansible_fqdn` as `localhost` (*).
(*): It could also be using a local fact for the fqdn?

Hypothesis to test: this did not happen pre-buster --> it could have something to do with python version / behaviour having changed for `socket.getfqdn`.

**Comment by drybjed** [comment]
The `debops.netbase` role [already has an entry for IPv6 address](https://github.com/debops/debops/blob/master/ansible/roles/debops.netbase/defaults/main.yml#L142-L143), does that show up in the `/etc/hosts` file? Can you show the output of the
```
ansible <host> -m setup -b
```
command on a fresh host in that environment so we can see what Ansible has to deal with?

I think that most interesting facts would be `ansible_default_ipv4.address` and `ansible_default_ipv6.address`. Can I assume that no IPv4 connectivity is intentional? In that case, perhaps the role shouldn't add the IPv4 address at all, so that double `127.0.0.1` entry doesn't happen. But I'm not sure how that looks from the Ansible side, does the default IPv4 address in facts is empty, or `127.0.0.1`? In the latter case that might be problematic.

**Comment by evilham** [comment]
Further testing now.

For completeness sake: the machine I discovered this on does have dual-stack networking, but only a public IPv6.
I do have some machines that have only and IPv6 stack (in case you are curious, these people at @ungleich are actually pretty awesome: http://ipv6onlyhosting.com/).

I just ran `debops bootstrap` on such an IPv6-only machine that I set up back in November, which is before this 7db8aa6cd2cbffb1429a578fcba7f0f6bad07b7f landed and I am seeing similar issues.

I think `debops.netbase` is not working as expected in this case :-D will figure out why and how to fix it.

Since you asked, this is for the IPv6-only machine:

````
"ansible_all_ipv4_addresses": [], 
"ansible_all_ipv6_addresses": [
    "2a0a:e5c0:X:X:X:X:X:X", 
    "fe80::X:X:X"
],
````

The `/etc/hosts` file looks like this in this case:
````
127.0.0.1	localhost
::1		localhost ip6-localhost ip6-loopback
fe00::0          ip6-localnet
fe00::0          ip6-mcastprefix
fe02::1          ip6-allnodes
fe02::1          ip6-allrouters
ff02::1		ip6-allnodes
ff02::2		ip6-allrouters
````

And this is for the dual-stack (but only public-facing IPv6) machine:

````
"ansible_all_ipv4_addresses": [
    "10.3.X.X"
], 
"ansible_all_ipv6_addresses": [
    "2001:470:X:X:X:X:X:X", 
    "fe80::X:X:X:X"
],
````

It may be useful to note that in both cases, the IPv6 addresses don't have `PTR` records, but they do have valid `AAAA` record for the `FQDN` that is setup in the ansible inventory).

**Comment by evilham** [comment]
This solves part of the problem:
https://github.com/debops/debops/blob/872a3584c720352555be91bef455ff9e075e25ee/ansible/roles/debops.netbase/defaults/main.yml#L70

That's exactly the behaviour I expected, may I ask why it's not the default? If there was a discussion about it, I totally missed it.

It doesn't *fully* fix the issue, since now `/etc/hosts` lacks a `::1 fqdn hostname` entry.

**Comment by drybjed** [comment]
The issue comes to having an environment where you have control over local DNS/DHCP services or not.

If you have that control, you can configure DNS records and DHCP host entries to whatever you want in one central location and you don't need to worry about individual `/etc/hosts` files. That's (I think) common in self-hosted or on-premise environments, where you manage your own infrastructure.

If you don't manage your own DNS and DHCP environment, then you might want to "take it over" one host at a time using `/etc/hosts` files. Fair enough, there's no better way to handle this in that case. Of course the entries in `/etc/hosts` will "mask" the entries coming from DNS; when that happens and the IP addresses that DNS points to change, you have to update your `/etc/hosts` files again... That's why we designed DNS in the first place to get rid of that problem.

Now, how would you distinguish between those two environments in an automated and foolproof way? In the first case, you want to leave the `/etc/hosts` database as it is and rely only on the DNS servers your hosts get via DHCP. In the second case, you have to manage `/etc/hosts` entries and you most likely want to ignore whatever the host (and Ansible) currently know about the domain so that you can set your own.

That's where the `netbase__domain` variable comes in. If you know that your environment is not under your full controll ahead of the time, you can set the domain you want in the `ansible/inventory/group_vars/all/netbase.yml` file, and the second behaviour will automatically be enabled, the role will configure your host's IP addresses with the hostname and domain you want. I think that's easiest solution to implement all of this right now; if you have better idea on how to distinguish both environments, let me know.

As for the IP address entries in `/etc/hosts`... You don't want an
```
::1 fqdn hostname
```
entry in the database, this will only confuse other services. Leave `::1` to `localhost`. What you want, is your default IPv6 address (`2001:470:...`) that points to the FQDN and the real hostname. This should already be added by the role, is it not? Also, the IPv4 address (`10.3...`) entry should also be present, in the dual stack mode. Does it show up in `/etc/hosts`?

The entries in the `/etc/hosts` file:
```
fe00::0          ip6-localnet
fe00::0          ip6-mcastprefix
fe02::1          ip6-allnodes
fe02::1          ip6-allrouters
```
seem to be anomalous, I think that these are added by the VPS provider in the OS image. Not really sure what to do about them, but I guess they shouldn't mess with anything in particular at this point.

Could you show the contents of the `/etc/hosts` files from the IPv6-only and dual-stack hosts after you apply the `debops.netbase` configuration on them? You will most likely have to set the `netbase__domain` variable for the correct results.

**Comment by evilham** [comment]
Oh, ffs, nevermind... This was way too odd for it not to be a typical problem between keyboard and chair.
Apparently things were in a weird state last time I pulled from master and I was seeing a different role form what was actually happening, I think I pulled right after 7db8aa6cd2cbffb1429a578fcba7f0f6bad07b7f but before 44435986a6c656d805e6dc901fd2b7df177b581d.

So all of this boils down to:
- Why does `netbase__domain` default to `''`? For example I fully expected the behaviour in https://github.com/debops/debops/blob/872a3584c720352555be91bef455ff9e075e25ee/ansible/roles/debops.netbase/defaults/main.yml#L70

It is kind of an issue, because that means that the defaults break a lot of things, e.g. after running `debops.netbase` the `ansible_fqdn` is not properly reported any longer, even if you had the fqdn properly setup before.

**Comment by evilham** [comment]
Oh, we cross-posted. I see the point about control over DNS / DHCP, it totally makes sense.
However, don't we already assume that the ansible inventory name *is* the FQDN?
Why would having `netbase__domain` default to "after the dot" be counterproductive?

I mean, my issue is now solved because I did override that as in the documentation, but in order to have the system fully recognise the FQDN, we need the `/etc/hosts` entry with either `::1` or the full IPv6 + `fqdn hostname`, regardless of DNS and DHCP.

**Comment by drybjed** [comment]
OK, so I wonder if having something like this as a default would be benefical in both cases:
```yaml
netbase__domain: '{{ "" if ansible_domain|d()
                        else (inventory_hostname.split(".")[1:] | join(".")) }}'
```
I'm currently checking what would be the value of `ansible_domain` in a freshly installed Debian box, can you tell me what's the value in your VPS? Also, when you set `netbase__domain`, does the role add the public IPv6 address in `/etc/hosts` on your VPS hosts like I mentioned earlier?

**Comment by evilham** [comment]
> I'm currently checking what would be the value of ansible_domain in a freshly installed Debian box, can you tell me what's the value in your VPS? Also, when you set netbase__domain, does the role add the public IPv6 address in /etc/hosts on your VPS hosts like I mentioned earlier?

Will give it a go as well. As for `netbase__domain`, setting it in the inventory adds a `/etc/hosts` entry with the public IPv6 + `fqdn hostname`.

**Comment by drybjed** [comment]
OK, so does that fix an empty `ansible_domain` and `ansible_fqdn` without a domain? Or, does Ansible correctly recognize the hostname, etc. from these entries on IPv6-only host? Also, other things like `ping` and `dig` and `getent hosts` work as expected?

**Comment by drybjed** [comment]
So I've checked what Debian Installer does, and it seems that even when a DNS/DHCP environment is present, `d-i` adds an entry to the `/etc/hosts` file like:
```
127.0.1.1    host.example.org host
```
Very well. I think that the `debops.netbase` role could deal with that by checking the conditions:

- is the FQDN real, it has dots or not

- does the FQDN point to `127.0.1.1` or not

Of course these checks need to happen from the remote host perspective and not Ansible Controller, so probably a `shell` task would be involved, or maybe even `getent` module, I'll check. Anyway, if the FQDN has a real domain in it and points to `127.0.1.1`, I would assume that there's a real DNS somewhere and the `127.0.1.1` entry can be safely removed and nothing added in its place. If the FQDN is not real and has only a hostname, we then can add real IP addresses and a domain in `/etc/hosts`. @evilham, what do you think about this solution?

**Comment by evilham** [comment]
I am also performing some checks, but I think I see an issue with what you are proposing, picture this:

- Hardware installed offline, maybe installed with a valid hostname but no domainname
- That means, `/etc/hosts` has a `127.0.0.1 hostname` line, but no fqdn is setup
- You then setup DNS properly, put the machine online and your inventory entry (fqdn in `debops_all_hosts`)
- Now you run `debops bootstrap --limit fqdn`
- That fqdn won't point to `127.0.0.1`, therefore the proposed solution would consider the fqdn to "not be real"
- It likely has issues on systems with IPv6-only network stack (waiting on a testing VM)
- `debops.netbase` will make sure there is a `127.0.0.1 hostname` in `/etc/hosts` but will ignore the fqdn

I thought it was highly recommended that the inventory has the `fqdn` of each host and in that case, we could just default `netbase__domain` to "anything after the dot and empty string if there are no dots".

This still allows people to use local domains as fqdn, just hostnames on the inventory, ...

Notice that having `netbase__domain` and `netbase__hostname` properly setup, results in the host being able to resolve the fqdn generated by those, regardless of the networking stack(s) used and it can be overridden by the users, while maybe being a saner default (I admit I may be overlooking cases, but I'd think that "fqdn in inventory" is a reasonable assumption and I can't find the bits of the documentation, but I think I read that somewhere on debops at some point).

**Comment by drybjed** [comment]
I'm working on a solution for IPv4 network, I'll make a PR in a bit and you can check if it works as expected with IPv6-only and dual-stack hosts.

Even if you install a host offline, without specifying a domain, Debian Installer will add an entry in `/etc/hosts` file:
```
127.0.1.1 hostname
```
I think that it would do this even in IPv6-only environment. Your VPS hosts are probably spun up from existing disk images, which might not have `127.0.1.1` entry set statically. Do you use something like `cloud-init` to set the hostname? The `/etc/hosts` contents you showed didn't have any hostname-related entries, so I would assume no.

**Comment by evilham** [comment]
They did, but my odd in-between commits pulled `debops.netbase` had gotten rid of `127.0.0.1 hostname`.
I will gladly help you test this :-). Thank you!

**Comment by drybjed** [comment]
@evilham A new PR with the fix is done (https://github.com/debops/debops/pull/622), can you check if that works as expected in both environments?

**Comment by evilham** [comment]
Sorry, I hadn't had the chance to check the merged solution and was relying on my overridden `netbase__domain`.

I just checked the changes and ran a few tests I had pending: results are good. Thank you!
